quinn-proto/src/congestion/bbr.rs

```diff
@@ -0,0 +1,653 @@
+        } else if (self.cwnd_gain < target_window as f32) || (self.acked_bytes < self.init_cwnd) {
+            // If the connection is not yet out of startup phase, do not decrease
+            // the window.
+            self.cwnd += bytes_acked;
+        }
+
+        // Enforce the limits on the congestion window.
+        if self.cwnd < self.min_cwnd {
+            self.cwnd = self.min_cwnd;
+        }
+    }
+
+    fn calculate_recovery_window(&mut self, bytes_acked: u64, bytes_lost: u64, in_flight: u64) {
+        if !self.recovery_state.in_recovery() {
+            return;
+        }
+        // Set up the initial recovery window.
+        if self.recovery_window == 0 {
+            self.recovery_window = self.min_cwnd.max(in_flight + bytes_acked);
+            return;
+        }
+
+        // Remove losses from the recovery window, while accounting for a potential
+        // integer underflow.
+        if self.recovery_window >= bytes_lost {
+            self.recovery_window -= bytes_lost;
+        } else {
+            const K_MAX_SEGMENT_SIZE: u64 = MAX_DATAGRAM_SIZE;
+            self.recovery_window = K_MAX_SEGMENT_SIZE;
+        }
+        // In CONSERVATION mode, just subtracting losses is sufficient.  In GROWTH,
+        // release additional |bytes_acked| to achieve a slow-start-like behavior.
+        if self.recovery_state == RecoveryState::Growth {
+            self.recovery_window += bytes_acked;
+        }
+
+        // Sanity checks.  Ensure that we always allow to send at least an MSS or
+        // |bytes_acked| in response, whichever is larger.
+        self.recovery_window = self
+            .recovery_window
+            .max(in_flight + bytes_acked)
+            .max(self.min_cwnd);
+    }
+
+    /// https://datatracker.ietf.org/doc/html/draft-cardwell-iccrg-bbr-congestion-control#section-4.3.2.2
+    fn check_if_full_bw_reached(&mut self, app_limited: bool) {
+        if app_limited {
+            return;
+        }
+        let target = (self.bw_at_last_round as f64 * K_STARTUP_GROWTH_TARGET as f64) as u64;
+        let bw = self.max_bandwidth.get_estimate();
+        if bw >= target {
+            self.bw_at_last_round = bw;
+            self.round_wo_bw_gain = 0;
+            self.ack_aggregation.max_ack_height.reset();
+            return;
+        }
+
+        self.round_wo_bw_gain += 1;
+        if self.round_wo_bw_gain >= K_ROUND_TRIPS_WITHOUT_GROWTH_BEFORE_EXITING_STARTUP as u64
+            || (self.recovery_state.in_recovery())
+        {
+            self.is_at_full_bandwidth = true;
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct LossState {
+    lost_bytes: u64,
+}
+
+impl LossState {
+    pub fn reset(&mut self) {
+        self.lost_bytes = 0;
+    }
+
+    pub fn has_losses(&self) -> bool {
+        self.lost_bytes != 0
+    }
+}
+
+/// A not so simple congestion controller to overcome high latency and packet loss
```

Comment:
Let's cite the code this is based on, as well as the current revisions of the papers documenting the algorithm. A brief summary of how it compares to Cubic (e.g. "aims for reduced buffer bloat and improved performance over high bandwidth-delay product networks", if that's accurate) would also be nice.

I'm also not sure any of us are currently able to review the *correctness* of this implementation, and afaik its development is still ongoing upstream, so maybe we should also clearly (i.e. in the first line) designate it as experimental.

Reply:
I think BBR is pretty much done, but BBR2 is under development

Reply:
changed description